### ISSUE TITLE
At normal mode, character under cursor is not inverse make me can not see it

### ISSUE THREAD

**Issue by gngshn** [neovim/neovim#6577]
- `nvim --version`:
NVIM v0.2.0-1451-g7f6d3d3
Build type: Release
Compilation: /usr/bin/cc -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -O2 -DNDEBUG -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/home/grant_shen/backup/neovim/build/config -I/home/grant_shen/backup/neovim/src -I/home/grant_shen/backup/neovim/.deps/usr/include -I/home/grant_shen/backup/neovim/.deps/usr/include -I/home/grant_shen/backup/neovim/.deps/usr/include -I/home/grant_shen/backup/neovim/.deps/usr/include -I/home/grant_shen/backup/neovim/.deps/usr/include -I/home/grant_shen/backup/neovim/.deps/usr/include -I/usr/include -I/home/grant_shen/backup/neovim/build/src/nvim/auto -I/home/grant_shen/backup/neovim/build/include
Compiled by grant_shen@grant-pc

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"
- Vim (version: ) behaves differently?
gvim does not have this problem
- Operating system/version:
Fedora 25
- Terminal name/version:
GNOME Terminal 3.22.1 Using VTE version 0.46.1 +GNUTLS
- `$TERM`:
xterm-256color

### Actual behaviour
**This issue is occurred after neovim support guicursor**. At normal mode, character under the cursor is not reverse, so it have the same color with cursor, then I can not see the character.
### Expected behaviour
color of character under cursor at normal mode need contrary to the cursor's color
### Steps to reproduce using `nvim -u NORC`
It's the same.
```
nvim -u NORC

```
example: When I type a `test` in neovim, and press esc back to normal mode. the `t` will be covered by the cursor because it does not change its color to `hi Cursor`
![guicursor](https://cloud.githubusercontent.com/assets/7463638/25367330/7fbb1194-293a-11e7-99f5-1f681f729654.png)

**Comment by teto** [comment]
the guifg attribute of the cursor highlight is not supported yet and I am not sure how easy it is as it may involve rewriting text as the cursor moves if no adequate CSI/OCS exists.
The mode that inverts text color and cursor color is I believe called "reverse video" (likely enabled by unibi_enter_reverse_mode). To configure it, the code could detect specialcase guibg=fg ?

**Comment by justinmk** [comment]
Ah, I see that http://invisible-island.net/xterm/ctlseqs/ctlseqs.html doesn't have a foreground cursor color sequence. [iTerm](https://www.iterm2.com/documentation-escape-codes.html) does (just `m` instead of `l`).

**Comment by justinmk** [comment]
> At normal mode, character under the cursor is not reverse, so it have the same color with cursor, then I can not see the character.

@gngshn Is that for all text, or only text whose color matches the `:hi Cursor` guibg?

**Comment by gngshn** [comment]
@justinmk 
If you try molokai(`tomasr/molokai`) theme(use termguicolors), you will know what I mean.
If the character has the same color with cursor(white), then when you move cursor on it. cursor will cover it. because the character does not change it color to `:hi Cursor`.

**Comment by teto** [comment]
reverse_mode does not seem to be concerned with cursor (`printf "\E[7mtoto\E[0m"` will output toto with colors inverted) so apart from writing current char with the cursor guifg and rewrite it when the cursor moves, which seems like a pain, I've no idea how to achieve that in the TUI. You can change the shape of the cursor or choose another cursor color that doesn't clash with common highlights.

**Comment by gngshn** [comment]
So is there only way to use terminal cursor instead of guicursor? Thinks

**Comment by jamessan** [comment]
`'guicursor'` just defines the configuration for the cursor in a more descriptive/portable way than requiring every user to look up terminal escape codes for their terminal emulator to plug-in to Vim's `t_SI`/`t_SR`/`t_EI` terminal options.

There may be some issues right now, but we're working to provide a cleaner way for the user to describe what's needed and centralizing the handling of the random terminals so user's _don't_ have to worry about that.

**Comment by gngshn** [comment]
@jamessan
Ok, thinks for your imformation

**Comment by jalvesaq** [comment]
I cannot reproduce the bug in Ubuntu 17.04 (GNOME Terminal 3.20.2 and XTerm 234). I created an empty `~/.config/nvim` directory and then an empty `colors` directory. I downloaded https://github.com/tomasr/molokai/blob/master/colors/molokai.vim into `~/.config/nvim/colors`. Then, I started Neovim as `nvim -u NORC`, typed random text and went to Normal mode. I also tested with a shell script. In XTerm, the cursor is always a block with light gray background and black foreground while in GNOME Terminal I get the syntax highlight reversed, that is, I can always see the character under the cursor. It is the same after `:colorscheme molokai`, and with Neovim inside Tmux.
